stan/math/prim/fun/grad_pFq.hpp

```diff
@@ -0,0 +1,340 @@
+ * \f$ \frac{\partial}{\partial a_1} =
+ *     \frac{z\prod_{j=2}^pa_j}{\prod_{j=1}^qb_j}
+ *     F_{q+1\:0\;1}^{p\:1\:2}\left(\begin{array}&a_1+1,...,a_p+1;1;1,a_1\\
+ *        2, b_1+1,...,b_1+1;;a_1+1\end{array};z,z\right) \f$
+ *
+ * \f$ \frac{\partial}{\partial b_1}=
+ *     -\frac{z\prod_{j=1}^pa_j}{b_1\prod_{j=1}^qb_j}
+ *     F_{q+1\:0\;1}^{p\:1\:2}\left(\begin{array}&a_1+1,...,a_p+1;1;1,b_1\\
+ *        2, b_1+1,...,b_1+1;;b_1+1\end{array};z,z\right) \f$
+ *
+ * \f$ \frac{\partial}{\partial z}= \frac{\prod_{j=1}^pa_j}{\prod_{j=1}^qb_j}
+ *      {}_pF_q(a_1 + 1,...,a_p + 1; b_1 +1,...,b_q+1;z) \f$
+ *
+ * @tparam calc_a Boolean for whether to calculate derivatives wrt to 'a'
+ * @tparam calc_b Boolean for whether to calculate derivatives wrt to 'b'
+ * @tparam calc_z Boolean for whether to calculate derivatives wrt to 'z'
+ * @tparam TupleT Type of tuple containing references
+ * @tparam Ta Type of a
+ * @tparam Tb Type of b
+ * @tparam Tz Type of z
+ * @param[in] grad_tuple Tuple of references to evaluate gradients into
+ * @param[in] a Vector of 'a' arguments to function
+ * @param[in] b Vector of 'b' arguments to function
+ * @param[in] z Scalar z argument
+ * @param[in] precision Convergence criteria for infinite sum
+ * @param[in] max_steps Maximum number of iterations for infinite sum
+ * @return Generalised hypergeometric function
+ */
+template <bool calc_a, bool calc_b, bool calc_z, typename TupleT, typename Ta,
+          typename Tb, typename Tz,
+          require_all_eigen_vector_t<Ta, Tb>* = nullptr,
+          require_stan_scalar_t<Tz>* = nullptr>
+void grad_pFq_impl(TupleT&& grad_tuple, const Ta& a, const Tb& b, const Tz& z,
+                   double precision = 1e-14, int max_steps = 1e6) {
+  using std::max;
+  using scalar_t = return_type_t<Ta, Tb, Tz>;
+  using Ta_plain = plain_type_t<Ta>;
+  using Tb_plain = plain_type_t<Tb>;
+  using T_vec = Eigen::Matrix<scalar_t, -1, 1>;
+
+  Ta_plain ap1 = (a.array() + 1).matrix();
+  Tb_plain bp1 = (b.array() + 1).matrix();
+  Ta_plain log_a = log(a);
+  Tb_plain log_b = log(b);
+  scalar_type_t<Tz> log_z = log(z);
+  scalar_type_t<Ta> sum_log_a = sum(log_a);
+  scalar_type_t<Tb> sum_log_b = sum(log_b);
+
+  // Only need the infinite sum for partials wrt p & b
+  if (calc_a || calc_b) {
+    double log_precision = log(precision);
+
+    int m = 0;
+    scalar_t outer_diff = 0;
+
+    // Declare vectors to accumulate sums into
+    // where NEGATIVE_INFTY is zero on the log scale
+    T_vec da_infsum = T_vec::Constant(a.size(), NEGATIVE_INFTY);
+    T_vec db_infsum = T_vec::Constant(b.size(), NEGATIVE_INFTY);
+
+    while ((outer_diff > log_precision) && (m < max_steps)) {
+      // Vectors to accumulate outer sum into
+      T_vec da_iter_m = T_vec::Constant(a.size(), NEGATIVE_INFTY);
+      T_vec da_mn = T_vec::Constant(a.size(), NEGATIVE_INFTY);
+      T_vec db_iter_m = T_vec::Constant(b.size(), NEGATIVE_INFTY);
+      T_vec db_mn = T_vec::Constant(b.size(), NEGATIVE_INFTY);
```

Comment:
You can make these outside of the loop and just reset them to the rhs within the loop